### DISCUSSION TITLE
Bagaimana cara agar file php di folder Asset hanya dapat diakses dengan tanpa ekstensi .php saja dan dapat menjalankan kondisi $_SERVER['REQUEST_METHOD'] ?

### DISCUSSION THREAD

**Discussion by bungfrangki** [mecha-cms/mecha#276] [Q&A] [answered]
saya membuat file json dengan _php_ di folder _asset_, sebut saja **memequote.php**
yang kemudian saja akan panggil dengan method POST pada file script.
Adapun file php tersebut saya tambahkan kondisi seperti ini:

``` php
<?php if ($_SERVER['REQUEST_METHOD'] === 'POST') {

} else if (stripos($_SERVER['REQUEST_URI'], basename(__FILE__)) !== FALSE) { 
    header('HTTP/1.0 403 ');
    die('<p>Access denied</p>');
}
```

nah, masalahnya file php ini dapat diakses dengan 2 cara, yaitu tanpa dan dengan ekstensi .php
ketika dibuka tanpa menggunakan ekstensi .php, maka kondisi di atas menjadi tidak berfungsi.

Lalu bagaimana caranya mas, biar file .php tersebut hanya bisa dibuka dengan tanpa ekstensi.php?
sekalian bisa memfungsikan kondisi di atas? thx

**Comment by taufik-nurrohman** [comment, marked as answer]
Ada mas caranya pakai ekstensi [mecha-cms/x.route](https://github.com/mecha-cms/x.route) :+1:

<https://mecha-cms.com/store/extension/route>

**Reply by bungfrangki** [reply]
mas, ekstensi Route ini basisnya hanya bekerja ditingkatan file yaa?
saya coba melakukan route langsung pada folder tapi gak bisa.
misalnya di folder ini: `/lot/asset/images/*`
idealnya ingin diakses seperti ini: `/images/*`

**Reply by taufik-nurrohman** [reply]
Tujuan utamanya untuk membuat rute virtual jadi nggak bisa kalau menyesuaikan struktur folder apa adanya.

Coba begini:

~~~ php
Hook::set('route', function ($content, $path) {
    if (!$path || 0 !== strpos($path, '/images/') {
        return $content;
    }
    $path = path(LOT . D . 'asset' . $path);
    if (is_file($path)) {
        test($path);
        exit;
    }
});
~~~

**Reply by taufik-nurrohman** [reply]
Oiya karena rute virtual jadi pengaturan _cache_ [harus manual](https://github.com/mecha-cms/x.feed/blob/e0980cd03f3df640f95ecd1d1e96cfe3f390456c/index.php#L353-L355) ya mas, ngeset _cache_ di `.htaccess` nggak akan memberikan pengaruh. Tanpa _cache_ bisa sebenarnya tapi performa web nanti jadi jelek.

**Reply by taufik-nurrohman** [reply]
Eh, tapi bisa ding mas, kalau cuma membuat berkas `.\lot\route\images.php`. Jalur sisa kan bisa diakses melalui variabel `$path` nanti:

~~~ php
return function ($content, $path) {
    $path = path(LOT . D . 'asset' . $path);
    if (is_file($path)) {
        test($path);
        exit;
    }
};
~~~

**Referensi:** https://github.com/mecha-cms/x.route/blob/579120acf93a218108bba67a1bd91cce4a080ea1/about.page#L55-L68

**Reply by bungfrangki** [reply]
outputnya ngambil `$path` yang seperti ini ya mas:
``` php
'C:\\wamp64\\www\\lot\\asset\\images\\logo.svg'
```

**Reply by taufik-nurrohman** [reply]
Ya nanti bisa dioper ke kelas `File`:

~~~ php
if (is_file($path)) {
    $file = new File($path);
    status(200, []);
    type($file->type);
    echo $file->content;
    exit;
}
~~~

**Reply by bungfrangki** [reply]
oh siyap.
saya berhasilnya seperti ini:
``` php
Hook::set('route', function ($content, $path) {
    if (!$path || 0 !== strpos($path, '/images/')) {
        return $content;
    };
    return content(LOT . D . 'asset' . D . $path);
});
```

**Reply by bungfrangki** [reply]
yg sy butuhkan sebetulnya adalah mengakses file setelahnya. tetapi sy ingin memotong jalur `/lot/asset/`
alhamdulillah berhasil dengan cara di atas. terima kasih mas (y)

**Reply by taufik-nurrohman** [reply]
Ya bisa. Tapi tipe MIME belum ditentukan jadi dia bakal coba nebak dari ekstensi berkas. Bisa sih.

**Comment by bungfrangki** [comment]
maaf mas, saya masih belum berhasil menerapkan extensi route di sisi saya. Saya masih bingung untuk nulis function-nya itu dimana?! di file php yg ingin kita rubah route-nya atau ada file khusus?

saat ini sy masih menggunakan setingan .htaccess seperti ini:
```
<Files ~ '[.](json|php)$'>
Allow from all
</Files>
```

untuk menghalau dualisme url pada file yg sama. Ya meski belum sesuai harapan, setidaknya pada file/url ber-ekstensi .php itu kondisi yang saya sebut di atas bisa berjalan dengan baik.

**Reply by taufik-nurrohman** [reply]
Buat berkas rute di folder `.\lot\route` akan secara otomatis menambahkan rute sesuai jalur berkas. Misal berkas `.\lot\route\random-quotes.php` akan bisa diakses di URL `http://127.0.0.1/random-quotes`

Di dalam berkas `random-quotes.php` diisi dengan konten seperti ini:

~~~ php
<?php

return function () {
    return 'Test quote...';
};
~~~

**Reply by taufik-nurrohman** [reply]
Berkas JSON bisa juga disimpan di folder yang sama. Nggak harus aksesibel secara publik si:

~~~ txt
random-quotes.json
random-quotes.php
~~~

~~~ php
<?php

return function () {
    $data = From::JSON(content(__DIR__ . D . 'random-quotes.json'), true);
    $randomIndex = /* ... */;
    return $data[$randomIndex];
};
~~~

**Comment by bungfrangki** [comment]
kok, keren si mas 😄🙏
ini berhasil, dan kondisinya juga bekerja dengan baik
Terima kasih mas atas solusinya
